plugins/gin/gin.go

```diff
@@ -33,11 +34,47 @@ const (
 	httpServerComponentID int32 = 49
 )
 
+type routeInfo struct {
+	operationName string
+}
+
+type middleware struct {
+	routeMap     map[string]map[string]routeInfo
+	routeMapOnce sync.Once
+}
+
 //Middleware gin middleware return HandlerFunc  with tracing.
-func Middleware(tracer *go2sky.Tracer) gin.HandlerFunc {
-	return func(c *gin.Context) {
+func Middleware(engine *gin.Engine, tracer *go2sky.Tracer) gin.HandlerFunc {
+	if engine == nil || tracer == nil {
+		return nil
+	}
+	m := new(middleware)
 
-		operationName := fmt.Sprintf("/%s%s", c.Request.Method, c.HandlerName())
+	return func(c *gin.Context) {
+		m.routeMapOnce.Do(func() {
```

Comment:
You might know that better than me but, isn't it better if we compute the list of routes previously and we pass that list to the `HandlerFunc` instead of computing every time we call it?

Reply:
hi, @jcchavezs   m.routeMapOnce.Do(func(){}) only compute once

Reply:
I still agree with @jcchavezs. Why can't this be computed before returning the handler? Simplifies code and eliminates the need for `sync.Once`.

Reply:
hi @basvanbeek  ,I tried to explain it, there is a logic problem here.
eg:

Scene1:
``` go

r := gin.New()
//Gin will use middleware before registering the route.
//so computing  before returning handler that routes will  always nil
r.Use(Middleware(r, tracer))
r.GET("/user/:name", func(c *gin.Context) {
		name := c.Param("name")
		c.String(200, "Hello %s", name)
	})
```
Scene2:
```go
r := gin.New()
r.GET("/user/:name", func(c *gin.Context) {
		name := c.Param("name")
		c.String(200, "Hello %s", name)
	})
//Gin will use middleware after registering the route. will find its.
r.Use(Middleware(r, tracer))
```
Scene 1 is standard and often used.

I don't know if you can understand my explanation.

Reply:
I understand. So at least make sure in all examples that it is best practice to add the go2sky middleware as your very last one, preferably right before calling `Run`. 

Reason: you want the go2sky middleware to be the first middleware to be hit (outer shell of the middleware onion) when serving a request (so you time most of the request, including time spent in other middleware).